src/kaleidoscope/MatrixAddr.h

```diff
@@ -0,0 +1,385 @@
+  }
+
+  template<typename MatrixAddr__>
+  ThisType& operator=(const MatrixAddr__ &other) {
+    raw_ = other.row() * rowSize + other.col();
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator+(const MatrixAddr__ &other) {
+    *this = ThisType(this->row_ + other.row(),
+                     this->col_ + other.col());
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator-(const MatrixAddr__ &other) {
+    *this = ThisType(this->row_ - other.row(),
+                     this->col_ - other.col());
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator+=(const MatrixAddr__ &other) {
+    *this = *this + other;
+    return *this;
+  }
+
+  template<typename MatrixAddr__>
+  ThisType &operator-=(const MatrixAddr__ &other) {
+    *this = *this - other;
+    return *this;
+  }
+};
+
+// Comparison operators only used for module testing. This is because they
+// cause ambiguous symbol lookup when used in the regular firmware.
+//
+// To use them also for the regular firmware they would need to be
+// disambiguated by moving them to the Kaleidoscope.h header and replacing
+// them with non-template versions that operate on the actual typedefed
+// KeyAddr and LEDAddr.
+
+#ifdef MATRIX_ADDR_TESTING
+
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator==(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  return (a1.row() == a2.row()) && (a1.col() == a2.col());
+}
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator!=(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  return !operator==(a1, a2);
+}
+
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator>(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  return (a1.row() > a2.row())
+         || ((a1.row() == a2.row()) && (a1.col() > a2.col()));
+}
+
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator<(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  // This could be optimized if necessary
+  return !operator>(a1, a2) && !operator==(a1, a2);
+}
+
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator>=(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  // This could be optimized if necessary
+  return operator>(a1, a2) || operator==(a1, a2);
+}
+
+template<typename MatrixAddr1__, typename MatrixAddr2__>
+bool operator<=(const MatrixAddr1__&a1, const MatrixAddr2__ &a2) {
+  return !operator>(a1, a2);
+}
+
+#endif
+
+} // end namespace kaleidoscope
+
+// Row/col based access functions have been superseded by matrix address
+// base access. To deprecate any row/col based access functions, toggle
+// the definitions of macro KS_ROW_COL_FUNC below
+// and uncomment _DEPRECATED_MESSAGE_ROW_COL_FUNC.
+//
+#define KS_ROW_COL_FUNC
+
+#if 0
+#define _DEPRECATED_MESSAGE_ROW_COL_FUNC \
+   "Row/col based access functions have been deprecated. Please use " \
+   "the KeyAttr/LEDAttr based versions instead."
```

Comment:
I believe you meant "KeyAddr/LEDAddr" (or "LedAddr") instead of "*Attr".

Reply:
Exactly.

Reply:
Fixed.